### PR TITLE
Add pdd.yandex.ru dns api support

### PR DESCRIPTION
**Pull request by non7top**
Add pdd.yandex.ru dns api support

### CODE REVIEW

File: dnsapi/dns_yandex.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# export PDD_Token="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+
+########  Public functions #####################
+
+#Usage: dns_myapi_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_yandex_add() {
+  fulldomain="${1}"
+  txtvalue="${2}"
+  _debug "Calling: dns_yandex_add() '${fulldomain}' '${txtvalue}'"
+  _PDD_credentials
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: please add code to check if the `_PDD_credentials()` function is success.

File: dnsapi/dns_yandex.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  export _H1="PddToken: $PDD_Token"
+
+  curDomain="$(echo "${fulldomain}" | awk -F. '{printf("%s.%s\n",$(NF-1), $NF)}')"
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: 1. do not use `awk`,  please use `sed` and `grep` instead.
2. do not split the root domain from the full domain.  try to use the yandex api to obtain the root domain list, then determine which part is a root domain.
see the `_get_root()` function here: https://github.com/Neilpang/acme.sh/blob/master/dnsapi/dns_cf.sh#L142

File: dnsapi/dns_yandex.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    _saveaccountconf PDD_Token "${PDD_Token}"
+  fi
+}
+
+pdd_get_record_id() {
+  fulldomain="${1}"
+  curDomain="$(echo "${fulldomain}" | awk -F. '{printf("%s.%s\n",$(NF-1), $NF)}')"
+  curUri="https://pddimp.yandex.ru/api2/admin/dns/list?domain=${curDomain}"
+  curResult="$(_get "${curUri}")"
+  echo "${curResult}" \
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: do not use python.  if you want to handle json data, please consider using `_normalizeJson()` , `_egrep_o()` , `sed` and `cut` instead.

File: dnsapi/dns_yandex.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    | python -c "
+import sys, json;
+rs=json.load(sys.stdin)[\"records\"]
+for r in rs:
+  if r[\"fqdn\"]==\"${fulldomain}\":
+    print r[\"record_id\"]
+    exit
+"
+}
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: add an empty line at the end of the file.

Comment (non7top) [inline dnsapi/dns_yandex.sh]: shfmt doesn't support this

File: dnsapi/dns_yandex.sh (hunk 2 above)
Comment (non7top) [inline dnsapi/dns_yandex.sh]: 1. removed awk
2. not sure about this, don't get the idea

File: dnsapi/dns_yandex.sh (hunk 1 above)
Comment (non7top) [inline dnsapi/dns_yandex.sh]: fixed

**Comment by non7top** [comment]
Ok, I tried to resolve the existing issues with the code. Please review.

File: dnsapi/dns_yandex.sh (hunk 2 above)
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: >> not sure about this, don't get the idea

Say you have a domain `mydomain.co.uk`,    you should not use the last 2 parts as the root domain.  Instead, you need to get a domain list from the yandex api to check which is the correct root domain.

File: dnsapi/dns_yandex.sh [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+pdd_get_record_id() {
+  fulldomain="${1}"
+  curDomain="$(echo "${fulldomain}" | rev | cut -d . -f 1-2 | rev)"
+  curSubdomain="$(echo "${fulldomain}" | rev | cut -d . -f 3- | rev)"
+  curUri="https://pddimp.yandex.ru/api2/admin/dns/list?domain=${curDomain}"
+  curResult="$(_get "${curUri}" | _normalizeJson)"
+  _debug "Result: $curResult"
+  echo "$curResult" | grep -o "{[^{]*\"content\":[^{]*\"subdomain\":\"${curSubdomain}\"" | sed -n -e 's#.* "record_id": \(.*\),[^,]*#\1#p'
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: do not `grep -o`,  some platforms doesn't support `-o` options.
please use `_egrep_o()` function instead.

File: dnsapi/dns_yandex.sh [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  curUri="https://pddimp.yandex.ru/api2/admin/dns/add"
+  curResult="$(_post "${curData}" "${curUri}")"
+  _debug "Result: $curResult"
+}
+
+#Usage: dns_myapi_rm   _acme-challenge.www.domain.com
+dns_yandex_rm() {
+  fulldomain="${1}"
+  _debug "Calling: dns_yandex_rm() '${fulldomain}'"
+  _PDD_credentials || exit 1
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: never use `exit`,  please use `return`

File: dnsapi/dns_yandex.sh [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+#Usage: dns_myapi_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_yandex_add() {
+  fulldomain="${1}"
+  txtvalue="${2}"
+  _debug "Calling: dns_yandex_add() '${fulldomain}' '${txtvalue}'"
+  _PDD_credentials || return 1
+  export _H1="PddToken: $PDD_Token"
+
+  curDomain="$(echo "${fulldomain}" | rev | cut -d . -f 1-2 | rev)"
```
Comment (Neilpang) [inline dnsapi/dns_yandex.sh]: do not cut the last 2 parts as the root domain.  Imaging you have a domain `mydomain.co.uk`.

You must get domain list from the yandex api, and check which part is the root domain.

**Comment by non7top** [comment]
I did a rework on domain detection.

**Comment by Neilpang** [comment]
please update the readme files:

https://github.com/Neilpang/acme.sh/blob/master/README.md
https://github.com/Neilpang/acme.sh/blob/master/dnsapi/README.md

**Comment by non7top** [comment]
Updated the documentation.

**Comment by Neilpang** [comment]
Please rebase your code,
It should be  No30.

**Comment by non7top** [comment]
not quite sure how rebase thing works, it somehow added some other commits in here

**Comment by non7top** [comment]
Thank you! How long does it often takes for changes from dev to be moved to master?

**Comment by Neilpang** [comment]
Merged.

**Pull request merged**